pkgs/by-name/sw/swaylock-plugin/package.nix

```diff
@@ -0,0 +1,80 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  meson,
+  ninja,
+  pkg-config,
+  scdoc,
+  libxcrypt,
+  wayland-scanner,
+  wayland,
+  wayland-protocols,
+  libxkbcommon,
+  cairo,
+  gdk-pixbuf,
+  pam,
+  nix-update-script,
+}:
+
+let
+  version = "1.8.0";
+  src = fetchFromGitHub {
+    owner = "mstoeckl";
+    repo = "swaylock-plugin";
+    rev = "v${version}";
+    hash = "sha256-Kd6Gqs+YnQu3qKfEeqW5CG38bU2gH2hqjoFEojWa8a4=";
+  };
+in
+stdenv.mkDerivation {
```

Comment:
What's the benefit of doing that?

Reply:
Thanks for the links.

This let binding shouldn't break overrideAttrs. I just double checked that with:

```
$ cat test.nix         
let
  pkgs = import ./. {};
in
  pkgs.swaylock-plugin.overrideAttrs(final: previous:{
    version = previous.version + "-hello";
  })
  
  (...)
  
$ nix-build test.nix            
this derivation will be built:
  /nix/store/by8mn2c6b9x26p1sm79q6aa8n1kn1gns-swaylock-plugin-1.8.0-hello.drv
building '/nix/store/by8mn2c6b9x26p1sm79q6aa8n1kn1gns-swaylock-plugin-1.8.0-hello.drv'...
Running phase: unpackPhase
```

(you can see the version number getting appended "hello" in the derivation name)

I was wondering what's the runtime cost of this let/in vs finalAttrs style:

For the let binding:
```json
{
  "cpuTime": 0.280131995677948,
  "envs": {
    "bytes": 1157624,
    "elements": 88356,
    "number": 56347
  },
  "gc": {
    "heapSize": 402915328,
    "totalBytes": 47709456
  },
  "list": {
    "bytes": 150776,
    "concats": 1604,
    "elements": 18847
  },
  "nrAvoided": 77783,
  "nrFunctionCalls": 47534,
  "nrLookups": 27758,
  "nrOpUpdateValuesCopied": 1675813,
  "nrOpUpdates": 5944,
  "nrPrimOpCalls": 39094,
  "nrThunks": 220785,
  "sets": {
    "bytes": 31040960,
    "elements": 1926584,
    "number": 13476
  },
  "sizes": {
    "Attr": 16,
    "Bindings": 16,
    "Env": 8,
    "Value": 24
  },
  "symbols": {
    "bytes": 267700,
    "number": 27576
  },
  "values": {
    "bytes": 9691704,
    "number": 403821
  }
}
```

For the finalAttrs fixed point:

```json

  "cpuTime": 0.23077300190925598,
  "envs": {
    "bytes": 1157648,
    "elements": 88358,
    "number": 56348
  },
  "gc": {
    "heapSize": 402915328,
    "totalBytes": 47709456
  },
  "list": {
    "bytes": 150776,
    "concats": 1604,
    "elements": 18847
  },
  "nrAvoided": 77782,
  "nrFunctionCalls": 47535,
  "nrLookups": 27758,
  "nrOpUpdateValuesCopied": 1675813,
  "nrOpUpdates": 5944,
  "nrPrimOpCalls": 39094,
  "nrThunks": 220788,
  "sets": {
    "bytes": 31040960,
    "elements": 1926584,
    "number": 13476
  },
  "sizes": {
    "Attr": 16,
    "Bindings": 16,
    "Env": 8,
    "Value": 24
  },
  "symbols": {
    "bytes": 267700,
    "number": 27576
  },
  "values": {
    "bytes": 9691776,
    "number": 403824
  }
}
```

The perf is overall pretty similar. The let/in version allocates two envs less, resulting in saving (1157624 bytes vs 1157648 bytes) 24 bytes.

Provided this small difference, I'd be up to keep the let/in style. However, if you prefer the fixed point style, I won't die on this hill, the perf difference is not big.

Reply:
> This is also different from `mkDerivation rec { ..... }`, which binds the recursive references immediately, so it allows you to reference original _inputs_ only.

^ reading this, I feel like I'm missing the point. But I cannot find an example where the let binding breaks the overrideAttrs system.

Reply:
Spent the last 30 minutes hitting my head on a wall reading `lib.extends`. I understand there's something I don't understand, but I can't figure out what kind of override this early let binding breaks. I'll likely figure that out at some point, but not tonight :D

I'll use the  `makeDerivationExtensible`  variant as you suggested above in the meantime.

Reply:
In case any bystander is reading this, @r-vdp gave me an example in which the early let binding is breaking things:

```
nix-repl> ((stdenv.mkDerivation (finalAttrs: { pname = "foo"; version = 1; src = fetchurl { url = "https://foo.org/${toString finalAttrs.version}"; }; })).overrideAttrs { version = 2; }).src.url
"https://foo.org/2"

nix-repl> ((let version = 1; in stdenv.mkDerivation { pname = "foo"; inherit version; src = fetchurl { url = "https://foo.org/${toString version}"; }; }).overrideAttrs { version = 2; }).src.url
"https://foo.org/1"
```

So overall, the derivation computation usually works, even with the early let binding. However, probing the derivation attributes **before** the full derivation evaluation breaks.